midterm_project/midterm-project - Eva - Ionut/muffin_man_shopping_list.py

```diff
@@ -0,0 +1,350 @@
+        pretty_string = '*' * number_of_stars + '\n' + self.name + '\n' + '*' * number_of_stars + '\n'
+
+        for index, (key, value) in enumerate(self.ingredients.items(), start=1):
+            pretty_string = '\n'.join((pretty_string, f'{index}. {key.title()}: {value}'))
+
+        pretty_string = pretty_string + '\n' * 2 + '*' * number_of_stars
+        return pretty_string
+
+    def __getitem__(self, given_index):
+        for index, (key, value) in enumerate(self.ingredients.items(), start=1):
+            if given_index == index:
+                return {key: value}
+
+    def __len__(self):
+        return len(self.ingredients)
+
+    def __iter__(self):
+        return iter(self.ingredients)
+
+    def keys(self):
+        return self.ingredients.keys()
+
+    @property
+    def name(self):
+        return self._name
+
+    @name.setter
+    def name(self, name):
+        if self.is_initialized:
+            raise NameError("You can't update the recipe!")
+        self._name = name
+
+    @property
+    def ingredients(self):
+        return self._ingredients
+
+    @ingredients.setter
+    def ingredients(self, ingredients):
+        if self.is_initialized:
+            raise NameError("You can't update the recipe!")
+        self._ingredients = ingredients
+        self.is_initialized = True
+
+
+class RecipesBox(object):
+
+    def __init__(self):
+        self.recipes_box = {}
+
+    def __getitem__(self, index):
+        return self.recipes_box[index]
+
+    def __delitem__(self, index):
+        del self.recipes_box[index]
+
+    def __setitem__(self, index, recipe):
+        self.recipes_box[index] = recipe
+
+    def __len__(self):
+        return len(self.recipes_box)
+
+    def insert(self, index, recipe):
+        self.recipes_box.insert(index, recipe)
+
+    def __iter__(self):
+        return iter(self.recipes_box)
+
+    def __str__(self):
+        return f'Recipes Box: {list(self.recipes_box.keys())}'
+
+    def add_recipe(self, recipe):
+        if recipe.name in self.recipes_box.keys():
+            print(f'Action Log: Recipe {recipe.name} was already in the Recipe Box')
+        else:
+            self.recipes_box[recipe.name] = recipe.ingredients
+            print(f'Action Log: Recipe {recipe.name} has been added to the Recipe Box')
+
+    def delete_recipe(self, recipe):
+        if recipe.name in self.recipes_box.keys():
+            del self.recipes_box[recipe.name]
+            print(f'Action Log: Recipe {recipe.name} has been removed from the Recipe Box')
+        else:
+            print(f'Action Log: Cannot remove recipe {recipe.name}: recipe not found in Recipe Box')
+
+    def pick_recipe(self, recipe=None):
+        if not recipe:
+            choices = list(self.recipes_box.keys())
+            recipe_choice = random.choice(choices)
+            print(f'Here is a random recipe {recipe_choice} and its ingredients {self.recipes_box[recipe_choice]}')
+        else:
```

Comment:
aici trebuie sa si verificati daca reteta respectiva exista in `recipes_box`. Daca trimiteti la pick_recipe o reteta care nu este deja salvata in recipes_box o sa fie un pic missleading pentru ca printu ii construit cu ajutorul `recipe` care este parametru si vine din exterior.